# Module Import Test Notebook

This notebook validates that the refactored modules load correctly.

In [1]:
#========================================
# Setup paths and notebook defaults
#========================================
from pathlib import Path
import sys

import numpy as np
import pandas as pd
import matplotlib as mpl
import matplotlib.pyplot as pp
import seaborn as sns

BASE_DIR = Path.cwd().resolve().parent
DATA_DIR = BASE_DIR / "data"
INPUT_DIR = DATA_DIR / "input"


if str(BASE_DIR) not in sys.path:
    sys.path.insert(0, str(BASE_DIR))

print(f"BASE_DIR: {BASE_DIR}")

BASE_DIR: C:\Users\F.Turner\Documents\MwM Proof of Concept V2


In [2]:
from src.ingest import *
from src.aggregate import *
from src.combine import *
from src.output import *

print("Imports succeeded: ingest.py, aggregate.py, combine.py, and output.py are available.")

Imports succeeded: ingest.py, aggregate.py, combine.py, and output.py are available.


In [3]:
df = pd.read_csv(DATA_DIR / "output" / "kenya" / "tables" / "combined_child_level.csv")

In [4]:
df

,date_time,admin1,admin2,admin3,school,sex,age,grade,age_group,haldo_sel,haldo_literacy,_uuid,location,protection_status,disability,round,aser_literacy,aser_numeracy
0,2023-06-12,8,800.0,80005.0,school01,1,7.0,0.0,6-10,0.604167,0.175000,1923914235876098121,NaN,NaN,NaN,NaN,NaN,NaN
1,2023-06-12,8,800.0,80005.0,school02,1,7.0,0.0,6-10,0.604167,0.175000,14316521119066352241,NaN,NaN,NaN,NaN,NaN,NaN
2,2023-06-07,9,902.0,90201.0,school03,1,7.0,2.0,6-10,0.837500,0.431818,8029136157369546520,NaN,NaN,NaN,NaN,NaN,NaN
3,2023-06-07,9,902.0,90201.0,school04,1,7.0,1.0,6-10,0.437500,0.431818,10855389975063662343,NaN,NaN,NaN,NaN,NaN,NaN
4,2023-06-08,9,902.0,90203.0,school05,1,8.0,0.0,6-10,0.500000,0.150000,17252436797090343061,NaN,NaN,NaN,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1461,2023-06-13,NaN,NaN,community4,school07,2,NaN,NaN,NaN,NaN,NaN,db9a6281-3825-4993-b9b3-76117e5fd0c0,NaN,NaN,NaN,NaN,NaN,NaN
1462,2023-06-13,NaN,NaN,community5,school08,3,NaN,NaN,NaN,NaN,NaN,07170e16-37fb-4a65-ae2b-fbe97d8a96e7,NaN,NaN,NaN,NaN,NaN,NaN
1463,2023-06-13,NaN,NaN,community5,school09,3,NaN,NaN,NaN,NaN,NaN,435e01cb-dc87-459e-b2da-bfd33087ef29,NaN,NaN,NaN,NaN,NaN,NaN
1464,2023-06-13,NaN,NaN,community5,school01,3,NaN,NaN,NaN,NaN,NaN,cf51583a-a512-4f2d-ac28-9b83162763d5,NaN,NaN,NaN,NaN,NaN,NaN


In [5]:
ds_obs = build_dataset(file_path=str(INPUT_DIR / "lesson_obs.xlsx"))
ds_obs = aggregates(ds_obs)
combined = add_to_combined(ds_obs, output_dir=BASE_DIR / "data" / "output")

KeyboardInterrupt: Interrupted by user

In [ ]:
ds_aser = build_dataset(file_path=str(INPUT_DIR / "aser_sample.csv"))
ds_aser = aggregates(ds_aser)
combined_aser = add_to_combined(ds_aser, output_dir=BASE_DIR / "data" / "output")

Form ID aser_001 already exists in the log. Processed data will be added to the existing dataset.
Metadata added to log:
{'form_id': 'aser_001', 'survey_name': 'aser_assessment', 'level': 'child', 'first_submission': '01/01/2026 00:00', 'last_submission': '03/01/2026 00:00', 'rows': 200, 'country': 'kenya', '_uuid_cols': 'child_uid,school_name,child_age'}
Data imported from file. Kobo metadata functions will be skipped (question dataframe and choices dataframe). Only basic metadata will be logged.
Dropped columns: ['child_name'] to protect child privacy
Using previously logged hash columns for UUID generation: ['child_uid', 'school_name', 'child_age']
Available dimensions: ['date_time', 'admin1', 'location', 'protection_status', 'school', 'sex', 'age', 'disability', 'round', 'age_group']
Missing dimensions: ['admin0', 'admin2', 'admin3', 'grade']
Resolved columns: {'date_time': '_submission_time', 'admin0': None, 'admin1': 'region', 'admin2': None, 'admin3': None, 'location': 'location

In [ ]:
ds_haldo = build_dataset(file_path=str(INPUT_DIR / "haldo_sample.xlsx"))
ds_haldo = aggregates(ds_haldo)
combined_haldo = add_to_combined(ds_haldo, output_dir=BASE_DIR / "data" / "output")

C:\Users\F.Turner\Documents\MwM Proof of Concept V2\src\ingest.py:286: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  country = None


Form ID haldo_001 already exists in the log. Processed data will be added to the existing dataset.
Metadata added to log:
{'form_id': 'haldo_001', 'survey_name': 'haldo_assessment', 'level': 'child', 'first_submission': '11/04/2023 11:40', 'last_submission': '13/06/2023 12:27', 'rows': 1246, 'country': 'kenya', '_uuid_cols': 'start,end,childid'}
Data imported from file. Kobo metadata functions will be skipped (question dataframe and choices dataframe). Only basic metadata will be logged.
Using previously logged hash columns for UUID generation: ['start', 'end', 'childid']
Available dimensions: ['date_time', 'admin1', 'admin2', 'admin3', 'school', 'sex', 'age', 'grade', 'age_group']
Missing dimensions: ['admin0', 'location', 'protection_status', 'disability', 'round']
Resolved columns: {'date_time': '_submission_time', 'admin0': None, 'admin1': 'governorate', 'admin2': 'district', 'admin3': 'subdistrict', 'location': None, 'protection_status': None, 'school': 'school', 'sex': 'sex', 'ag

In [ ]:
ds_school = build_dataset(file_path=str(INPUT_DIR / "school_sample.csv"))
ds_school = aggregates(ds_school)
ds_school_combined = school_combined(ds_school, output_dir=BASE_DIR / "data" / "output")

Form ID school_001 added to log. Processed data will be saved as a new file.
Metadata added to log:
{'form_id': 'school_001', 'survey_name': 'school_summary', 'level': 'school', 'first_submission': Timestamp('2026-01-01 00:00:00'), 'last_submission': Timestamp('2026-03-01 00:00:00'), 'rows': 27, 'country': 'kenya', '_uuid_cols': None}
Data imported from file. Kobo metadata functions will be skipped (question dataframe and choices dataframe). Only basic metadata will be logged.
UID depends on 7 columns - may be unstable
Available dimensions: ['date_time', 'admin1', 'admin2', 'school']
Missing dimensions: ['admin0', 'admin3', 'location', 'protection_status', 'sex', 'age', 'disability', 'grade', 'round', 'age_group']
Resolved columns: {'date_time': '_submission_time', 'admin0': None, 'admin1': 'region', 'admin2': 'district', 'admin3': None, 'location': None, 'protection_status': None, 'school': 'school_name', 'sex': None, 'age': None, 'disability': None, 'grade': None, 'round': None, 'age

C:\Users\F.Turner\Documents\MwM Proof of Concept V2\src\combine.py:251: UserWarning: Converting to PeriodArray/Index representation will drop timezone information.
  .groupby(["school", teacher_combined["date_time"].dt.to_period("M")], dropna=False)


In [ ]:
ds_school_combined.df_short.head(30)

,school,date_time,admin1,admin2,pta_active,pta_active,school_management_committee_active,plc_frequency,plc_frequency,safeguarding_policy,...,disruption_level,literacy_practice,numeracy_practice,sel_practice,student_teacher_ratio,learning_resources,gender_responsiveness,safe_environment,rtl_use,lesson_preparedness
0,school01,NaT,West,Zeta,no,no,yes,1.0,1.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,school01,2024-10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,3.000000,1.666667,2.0,1.666667,3.000000,2.333333,2.333333,2.666667,2.333333,2.333333
2,school01,2026-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,school02,NaT,West,Delta,no,no,yes,1.0,1.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,school02,2024-10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.333333,2.333333,3.0,2.333333,2.666667,1.666667,2.666667,1.666667,2.333333,2.666667
5,school02,2026-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
6,school03,NaT,East,Gamma,yes,yes,no,1.0,1.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
7,school03,2024-10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.500000,1.500000,3.0,2.000000,1.000000,2.000000,2.500000,3.000000,3.000000,2.000000
8,school03,2026-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
9,school04,NaT,South,Alpha,yes,yes,yes,1.0,1.0,1.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


Targeted Optimisation Plan

Cache config classification lookups (highest ROI, lowest risk)
Problem: each chart pair calls classification twice, and each call re-opens JSON files in report.py:55 and report.py:57, triggered from app.py:94 and app.py:95.
Change: load config once per level via cached helper (for example with st.cache_data) and classify using in-memory dicts.
Impact: removes repeated disk I/O across every rerun and every indicator-dimension pair.

Render only the currently selected level, not all three
Problem: all three tabs execute render paths in app.py:131, so one filter change recalculates Child + Teacher + School.
Change: replace level tabs with a level selector, or gate computation so only the active level runs.
Impact: often cuts dashboard compute by about 2/3 on each interaction.

Add a chart cap per indicator and lazy detail expansion
Problem: nested loops in app.py:83 and app.py:87 can generate many figures.
Change: show top N dimensions by default (for example 4-6), with an expander for “show all dimensions”.
Impact: large drop in Plotly construction/serialization time while preserving full detail on demand.

Precompute filtered dimension options once per rerun
Problem: option extraction and sorting in app.py:75 runs for each dimension every rerun.
Change: build an options map once from base df (or cached by country/level), then reuse in multiselects.
Impact: improves sidebar responsiveness, especially with high-cardinality columns.

Avoid repeated dataframe copies in hot loops
Problem: repeated slicing/copy/dropna in app.py:90 and conversion blocks around app.py:98 and app.py:116.
Change: do type normalization once per level before chart loops, and avoid unnecessary copy() unless mutation is required.
Impact: reduces memory churn and Python overhead.

Cache figure-ready aggregates for common plot paths
Problem: repeated groupby and crosstab in report.py:192, report.py:199, report.py:219.
Change: cache grouped outputs keyed by (country, level, indicator, dimension, filter_signature).
Impact: big gains when users tweak a subset of filters repeatedly.

Reduce per-figure axis/tick work
Problem: tick text rebuilding for every figure in report.py:251.
Change: only apply custom tick formatting for categorical axes with large labels; skip for numeric/clean short labels.
Impact: modest but noticeable when many charts are shown.

Keep data loading cached but tighten cache keys
Good: app.py:32 already caches load.
Improve: ensure cache depends only on (level, country) and clear cache only after ingestion (already done at app.py:184 and app.py:228).
Impact: avoids accidental unnecessary reloads.

Suggested rollout order (minimal structural change)

- Config caching in classification.
- Single active level rendering.
- Chart cap + optional “show all”.
- Loop/dataframe micro-optimizations.
- Optional aggregate caching if still slow.

How to validate each step

Add timing logs around app.py:54 and inside chart loop near app.py:87.
Track:
total rerun time
number of charts rendered
average time per chart
Compare baseline vs after each step, keeping same country/filter state.
If you want, I can draft a concrete patch plan with exact function signatures and cache-key structure before any code changes.

# **SUPABASE**

In [ ]:
pip install supabase

  Using cached httpx-0.28.1-py3-none-any.whl.metadata (7.1 kB)
  Using cached httpcore-1.0.9-py3-none-any.whl.metadata (21 kB)
  Using cached pycparser-3.0-py3-none-any.whl.metadata (8.2 kB)
Using cached httpx-0.28.1-py3-none-any.whl (73 kB)
Using cached httpcore-1.0.9-py3-none-any.whl (78 kB)
   ---------------------------------------- 0.0/2.0 MB ? eta -:--:--
   -------------------- ------------------- 1.0/2.0 MB 6.0 MB/s eta 0:00:01
   ---------------------------------------- 2.0/2.0 MB 5.9 MB/s  0:00:00
   ---------------------------------------- 0.0/3.8 MB ? eta -:--:--
   ---------- ----------------------------- 1.0/3.8 MB 6.1 MB/s eta 0:00:01
   ------------------------ --------------- 2.4/3.8 MB 6.1 MB/s eta 0:00:01
   -------------------------------------- - 3.7/3.8 MB 6.1 MB/s eta 0:00:01
   ---------------------------------------- 3.8/3.8 MB 6.0 MB/s  0:00:00
Using cached pycparser-3.0-py3-none-any.whl (48 kB)

  Attempting uninstall: websockets

    Found existing installat


[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [3]:
import os
from supabase import create_client, Client

In [4]:
from dotenv import load_dotenv

load_dotenv()

True

In [5]:

url: str = os.environ.get("SUPABASE_URL")
key: str = os.environ.get("SUPABASE_KEY")
supabase: Client = create_client(url, key)

In [6]:
from supabase import create_client
from dotenv import load_dotenv
import os

load_dotenv()

supabase = create_client(
os.getenv("SUPABASE_URL"),
os.getenv("SUPABASE_KEY")
)

print("Connected")

Connected


In [7]:
import requests

requests.get("https://www.google.com")

<Response [200]>

In [24]:
pip install python-certifi-win32

  Using cached python_certifi_win32-1.6.1-py2.py3-none-any.whl.metadata (1.4 kB)
Using cached python_certifi_win32-1.6.1-py2.py3-none-any.whl (7.3 kB)

   ---------- ----------------------------- 1/4 [vcs-versioning]
   ---------- ----------------------------- 1/4 [vcs-versioning]
   ---------- ----------------------------- 1/4 [vcs-versioning]
   ---------- ----------------------------- 1/4 [vcs-versioning]
   -------------------- ------------------- 2/4 [setuptools-scm]
   ---------------------------------------- 4/4 [python-certifi-win32]

Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: python.exe -m pip install --upgrade pip


In [14]:
from supabase import create_client

supabase = create_client(
url,
key
)


result = supabase.table("combined_school").select("*").limit(30).execute()
    
print(result)

data=[{'school': 'school01', 'date_time': None, 'admin1': 'West', 'admin2': 'Zeta', 'pta_active': 'no', 'pta_active_1': 'no', 'school_management_committee_active': 'yes', 'plc_frequency': 1, 'plc_frequency_1': 1, 'safeguarding_policy': 0, 'safeguarding_policy_1': 0, 'haldo_sel': 1.3571987864254, 'haldo_literacy': 0.447432962720732, 'aser_literacy': None, 'aser_numeracy': None, 'time_on_task': None, 'class_management_time': None, 'facilitation_time': None, 'off_task_behaviour': None, 'disruption_level': None, 'literacy_practice': None, 'numeracy_practice': None, 'sel_practice': None, 'student_teacher_ratio': None, 'learning_resources': None, 'gender_responsiveness': None, 'safe_environment': None, 'rtl_use': None, 'lesson_preparedness': None, 'uuid': '491024dc-2b93-4799-a078-53e8a0e8008e'}, {'school': 'school01', 'date_time': '2024-10', 'admin1': None, 'admin2': None, 'pta_active': None, 'pta_active_1': None, 'school_management_committee_active': None, 'plc_frequency': None, 'plc_freque

In [16]:
import pandas as pd

df = pd.DataFrame(result.data)

In [17]:
df.head()

,school,date_time,admin1,admin2,pta_active,pta_active_1,school_management_committee_active,plc_frequency,plc_frequency_1,safeguarding_policy,...,literacy_practice,numeracy_practice,sel_practice,student_teacher_ratio,learning_resources,gender_responsiveness,safe_environment,rtl_use,lesson_preparedness,uuid
0,school01,NaN,West,Zeta,no,no,yes,1.0,1.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,491024dc-2b93-4799-a078-53e8a0e8008e
1,school01,2024-10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,1.6666666666666667,2.0,1.6666666666666667,3.0,2.3333333333333335,2.3333333333333335,2.6666666666666665,2.3333333333333335,2.3333333333333335,cec8ecde-f670-43eb-9d2f-1cfc897503c8
2,school01,2026-01,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2faaa6c8-728f-46f0-a0f3-fff97b65b359
3,school02,NaN,West,Delta,no,no,yes,1.0,1.0,0.0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,a98ad417-8025-40ee-a8c2-e50eee14e542
4,school02,2024-10,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,2.3333333333333335,3.0,2.3333333333333335,2.6666666666666665,1.6666666666666667,2.6666666666666665,1.6666666666666667,2.3333333333333335,2.6666666666666665,2b506254-b006-4047-9d54-978a4c895cc7


In [35]:
df = pd.read_csv('C:\\Users\\F.Turner\\Documents\\MwM Proof of Concept V2\\data\\import_log.csv')



In [41]:
import numpy as np

In [42]:
df = df.replace({np.nan: None})

In [53]:
records = df.to_dict(
orient="records"
)

In [54]:
print(records[0])

{'form_id': 'aser_001', 'survey_name': 'aser_assessment', 'level': 'child', 'first_submission': '01/01/2026 00:00', 'last_submission': '03/01/2026 00:00', 'rows': 200, 'country': 'kenya', '_uuid_cols': 'child_uid,school_name,child_age', 'logged_at': None}


In [55]:

supabase.table(
    "import_log"
).insert(
    records
).execute()

APIResponse(data=[{'id': None, 'form_id': 'aser_001', 'survey_name': 'aser_assessment', 'level': 'child', 'country': 'kenya', 'records_imported': None, 'first_submission': '01/01/2026 00:00', 'last_submission': '03/01/2026 00:00', 'logged_at': None, '_uuid_cols': 'child_uid,school_name,child_age', 'rows': 200, '_uuid': '95389960-9c16-43cf-b753-cba445e34bac'}, {'id': None, 'form_id': 'aser_001', 'survey_name': 'aser_assessment', 'level': 'child', 'country': 'kenya', 'records_imported': None, 'first_submission': '01/01/2026 00:00', 'last_submission': '03/01/2026 00:00', 'logged_at': None, '_uuid_cols': 'child_uid,school_name,child_age', 'rows': 200, '_uuid': '21176788-ba21-4c5a-9739-f8c1752b4e95'}, {'id': None, 'form_id': 'haldo_001', 'survey_name': 'haldo_assessment', 'level': 'child', 'country': 'kenya', 'records_imported': None, 'first_submission': '11/04/2023 11:40', 'last_submission': '13/06/2023 12:27', 'logged_at': None, '_uuid_cols': 'start,end,childid', 'rows': 1246, '_uuid': '4

In [56]:
indicators = pd.read_excel(r"C:\Users\F.Turner\Documents\MwM Proof of Concept V2\config\indicator_config.xlsx", sheet_name="Indicators")
calculations = pd.read_excel(r"C:\Users\F.Turner\Documents\MwM Proof of Concept V2\config\indicator_config.xlsx", sheet_name="Calculations")
components = pd.read_excel(r"C:\Users\F.Turner\Documents\MwM Proof of Concept V2\config\indicator_config.xlsx", sheet_name="Components")
domains = pd.read_excel(r"C:\Users\F.Turner\Documents\MwM Proof of Concept V2\config\indicator_config.xlsx", sheet_name="Domains")

In [72]:
dimensions = pd.read_excel(r"C:\Users\F.Turner\Documents\MwM Proof of Concept V2\config\Dimensions_config.xlsx", sheet_name="Dimensions")
columns = pd.read_excel(r"C:\Users\F.Turner\Documents\MwM Proof of Concept V2\config\Dimensions_config.xlsx", sheet_name="Columns")
values = pd.read_excel(r"C:\Users\F.Turner\Documents\MwM Proof of Concept V2\config\Dimensions_config.xlsx", sheet_name="Values")
operations = pd.read_excel(r"C:\Users\F.Turner\Documents\MwM Proof of Concept V2\config\Dimensions_config.xlsx", sheet_name="Operations")
bins = pd.read_excel(r"C:\Users\F.Turner\Documents\MwM Proof of Concept V2\config\Dimensions_config.xlsx", sheet_name="Bins")


In [76]:
tables = {
    "indicator_indicators": indicators, 
          "indicator_calculations": calculations, 
          "indicator_components": components, 
        #   "domains": domains,
          "dimensions": dimensions,
          "dimension_columns": columns,
          "dimension_values": values,
          "dimension_operations": operations,
          "dimension_bins": bins
          } 

for table_name, t in tables.items():
    t = t.replace({np.nan: None})
    records = t.to_dict(orient="records")
    print(table_name)
    print(len(records))
    print(records[:2])
    supabase.table(
        table_name
    ).upsert(
        records
    ).execute()

indicator_indicators
28
[{'id': 'haldo_self_concept', 'label': 'Haldo Self Concept', 'description': None, 'level': 'child', 'indicator_type': 'numeric', 'aggregation_type': 'calculated', 'aggregation_method': 'mean', 'missing_handling': 'ignore', 'active': True}, {'id': 'haldo_empathy', 'label': 'Haldo Empathy', 'description': None, 'level': 'child', 'indicator_type': 'numeric', 'aggregation_type': 'calculated', 'aggregation_method': 'mean', 'missing_handling': 'ignore', 'active': True}]
indicator_calculations
10
[{'indicator_id': 'haldo_self_concept', 'operation': 'mean', 'divide_by': None, 'rounding': None, 'notes': None}, {'indicator_id': 'haldo_empathy', 'operation': 'mean', 'divide_by': None, 'rounding': None, 'notes': None}]
indicator_components
36
[{'indicator_id': 'haldo_self_concept', 'component_name': 'selname', 'sort_order': 1}, {'indicator_id': 'haldo_self_concept', 'component_name': 'selage', 'sort_order': 2}]
dimensions
14
[{'id': 'date_time', 'dimension_type': 'ordinal',

APIError: {'message': 'permission denied for table dimensions', 'code': '42501', 'hint': 'Grant the required privileges to the current role with: GRANT SELECT, INSERT, UPDATE ON public.dimensions TO service_role;', 'details': None}